<a href="https://colab.research.google.com/github/Yamito548/Trabajos-cursos-IA-y-big-data/blob/main/act_3_5_calidad_vinos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

En este trabajo usaremos árboles de decisión para predecir la calidad de los vinos. Los dataset se descargaron [aquí](https://archive.ics.uci.edu/dataset/186/wine+quality).

# Importamos el dataset.

In [6]:
#(5%) Importación de los datasets (utilizar el dataset winequality-red.csv)
import pandas as pd
data = pd.read_csv("winequality-red.csv")
data

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"
0      7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                     
1      7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5                                                                                                                     
2     7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;...                                                                                                                     
3     11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58...                                                                                                                     
4      7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                     
...                                                 ...                                                                                                                     
1594  6.2;0.6;0.08;2;0.09;32;44;0.9949;3.45;0.58;10.5;5                                                                                                                     
1595  5.9;0.55;0.1;2.2;0.062;39;51;0.99512;3.52;0.76...                                                                                                                     
1596  6.3;0.51;0.13;2.3;0.076;29;40;0.99574;3.42;0.7...                                                                                                                     
1597  5.9;0.645;0.12;2;0.075;32;44;0.99547;3.57;0.71...                                                                                                                     
1598  6;0.31;0.47;3.6;0.067;18;42;0.99549;3.39;0.66;...                                                                                                                     

[1599 rows x 1 columns]

Podemos ver que hay 1599 filas, y sola una columna (en  realidad 12 separadas) en donde están todas las características juntas separadas por “;”.Todas las características están relacionadas a los vinos en si desde su acidez hasta su pH. Pero igual tendré que buscar un método para separar las características.

# Otra forma de cargar los datos (solución al problema anterior)

In [7]:
from google.colab import files
import io
def upload_files ():
  uploaded = files.upload()
  for fn in uploaded.keys():
    print('User uploaded file "{name}" with length {length} bytes'.format(
        name=fn, length=len(uploaded[fn])))
    df = pd.read_csv(io.StringIO(uploaded[fn].decode('utf-8')),sep=";")
    return df
#Al principio no me funciona bien pero al añadir sep=";" ya funcionó correctamente

In [ ]:
data = upload_files()
data

# Matriz de correlación

In [ ]:
#(5%) Mostrar la matriz de correlación de variables.
import seaborn as sns
import matplotlib.pyplot as plt
matriz_corr = data.corr()
sns.set(style='white')
plt.figure(figsize=(10,8))
sns.heatmap(matriz_corr, annot=True, cmap='coolwarm', fmt=".2f", linewidths=.5)

Aquí podremos ver como se relación todas las características entre ellas. A primera vista vemos que no hay ninguna característica la cual se lleve súper bien con otra pero veo buenos números igualmente como 0.67 que a primera vista es el valor más alto de correlación. Hay que recordar que para esta matriz los valores que nos muestran estan entre -1 y 1, siendo -1 una muy mala correlación y 1 una muy buena.

# Selección de características

kbest

In [31]:
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
#Dividimos los datos
X = data.drop("quality",axis=1)
y = data["quality"]
modelo = GaussianNB()
#Vemos los resultados que nos da con diferentes cantidades de características.

for n in range(2,5):
  k_best = SelectKBest(score_func=f_classif, k=n)
  X_best = k_best.fit_transform(X,y)
  select_column = X.columns[k_best.get_support()]
  X_train, X_test, y_train, y_test = train_test_split(X_best, y, test_size=0.20, random_state=100)
  modelo.fit(X_train,y_train)
  y_pred = modelo.predict(X_test)
  accuracy = accuracy_score(y_test, y_pred)
  print(f"Las características seleccionadas para {n} son: {select_column}")
  print(f"La precisión es de {accuracy}")

Las características seleccionadas para 2 son: Index(['volatile acidity', 'alcohol'], dtype='object')
La precisión es de 0.58125
Las características seleccionadas para 3 son: Index(['volatile acidity', 'total sulfur dioxide', 'alcohol'], dtype='object')
La precisión es de 0.6375
Las características seleccionadas para 4 son: Index(['volatile acidity', 'total sulfur dioxide', 'sulphates', 'alcohol'], dtype='object')
La precisión es de 0.621875


En este caso, la mayor precisión nos la da cuando solo hay 3 características las cuales son volatile acidity, total sulfur dioxide y alcohol dandonos un porcentaje de 0.63%. Luego de esto, a más características añadamos peores resultados nos va a dar.